src/theory/program_logic/total_program_logicScript.sml

```diff
@@ -141,56 +141,188 @@ QED
+    (*********************************************************)
+    qexists_tac `min_wf` >>
+    CONJ_TAC >| [
+      Q.UNABBREV_TAC `min_wf` >>
+      Q.UNABBREV_TAC `B''` >>
+      Q.UNABBREV_TAC `B'` >>
+      imp_res_tac pred_setTheory.CHOICE_DEF >>
+      gs[pred_setTheory.IN_ABS, pred_setTheory.IN_APP],
+
+      rpt strip_tac >>
+      PairCases_on `b` >>
+      PairCases_on `min_wf` >>
+      FULL_SIMP_TAC std_ss [] >>
+      Q.UNABBREV_TAC `B''` >>
+      subgoal `min_wf3 min_wf1 = x` >- (
+        fs[] >>
+        imp_res_tac pred_setTheory.CHOICE_DEF >>
+        gs[] >>
+        Q.UNABBREV_TAC `get_var` >>
+        Q.UNABBREV_TAC `get_state` >>
+        fs[]
+      ) >>
+      subgoal `get_wf_rel w = min_wf2` >- (
+        fs[] >>
+        imp_res_tac pred_setTheory.CHOICE_DEF >>
+        gs[] >>
+        Q.UNABBREV_TAC `get_wf_rel` >>
+        Q.UNABBREV_TAC `B'` >>
+        fs[]
+      ) >>
+      RW_TAC std_ss [] >>
+      QSPECL_X_ASSUM ``!y. get_wf_rel w y (min_wf3 min_wf1) ==> ~A' y`` [`(b3 b1)`] >>
+      REV_FULL_SIMP_TAC std_ss [] >>
+      Q.UNABBREV_TAC `A'` >>
+      Q.UNABBREV_TAC `B'` >>
+      FULL_SIMP_TAC std_ss [] >>
+      `IMAGE (\a. get_var a (get_state a))
+	     (\b. b IN (\b. b IN B /\ (get_wf_rel b = get_wf_rel w))) (b3 b1)` suffices_by (
+	FULL_SIMP_TAC std_ss []
+      ) >>
+      FULL_SIMP_TAC std_ss [pred_setTheory.IMAGE_applied] >>
+      qexists_tac `(b0,b1,get_wf_rel w,b3,b4,b5,b6,b7)` >>
+      Q.UNABBREV_TAC `get_var` >>
+      Q.UNABBREV_TAC `get_state` >>
+      Q.UNABBREV_TAC `get_wf_rel` >>
+      fs [pred_setTheory.IN_ABS, pred_setTheory.IN_APP]
+    ]
+  ]
+) >>
+rpt strip_tac >>
+fs [LET_DEF] >>
+IMP_RES_TAC pred_setTheory.CHOICE_DEF >>
+fs [GSYM pred_setTheory.MEMBER_NOT_EMPTY, pred_setTheory.IN_ABS]
 );
 
 
 Definition total_loop_jgmt_def:
- total_loop_jgmt TS l le invariant C1 var =
-  (~(l IN le) /\
-   !x:num. t_jgmt TS l ({l} UNION le) (\s. invariant s /\ C1 s /\ var s = x)
-        (\s. TS.ctrl s = l /\ invariant s /\ var s < x /\ var s >= 0))
+ total_loop_jgmt TS l le invariant C1 wf_rel var =
+  (l NOTIN le /\ WF wf_rel /\
+   !x. t_jgmt TS l ({l} UNION le) (\s. invariant s /\ C1 s /\ var s = x)
+        (\s. TS.ctrl s = l /\ invariant s /\ wf_rel (var s) x))
 End
 
 (* Note that due to loop_fun_ind relating states s and s' at loop points,
  * s needs to be exposed also here. Either this can be explicitly specified
  * in the precondition of the conclusion, or the definition can be unfolded, like here *)
 val loop_fun_ind_spec =
- Q.SPEC `\TS s var l le invariant C1.
+ Q.SPEC `\TS s wf_rel var l le invariant C1.
           first_enc TS ==>
-	  total_loop_jgmt TS l le invariant C1 var ==>
+          WF wf_rel ==>
+	  total_loop_jgmt TS l le invariant C1 wf_rel var ==>
 	  t_jgmt TS l le (\s. invariant s /\ ~C1 s) post ==>
-	  (invariant s /\ TS.ctrl s = l) ==>
+	  (invariant s /\ TS.ctrl s = l /\ C1 s) ==> (* TODO: Why does this need "C1 s" as conjunct also? *)
```

Comment:
I'm curious why earlier `C1 s` had not been needed for the loop invariant `loop_fun_ind_spec`.

Reply:
Also curious. Tried removing it - no apparent harm done with minimal changes needed to compensate. No idea how it appeared there.